common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -656,6 +771,206 @@ public void registerExecutor(String appId, ExecutorShuffleInfo executorInfo) {
+      return null;
+    }
+  }
+
+  /**
+   * Generate the DB key from AppAttemptShuffleMergeId object
+   */
+  private byte[] getDbAppAttemptShufflePartitionKey(
+      AppAttemptShuffleMergeId appAttemptShuffleMergeId) {
+    return getDbKey(appAttemptShuffleMergeId, APP_ATTEMPT_SHUFFLE_FINALIZE_STATUS_KEY_PREFIX);
+  }
+
+  /**
+   * Generate the DB key from AppAttemptId object
+   */
+  private byte[] getDbAppAttemptPathsKey(AppAttemptId appAttemptId){
+    return getDbKey(appAttemptId, APP_ATTEMPT_PATH_KEY_PREFIX);
+  }
+
+  /**
+   * Reload the DB to recover the meta data stored in the hashmap for merged shuffles.
+   * The application attempts local paths information will be firstly reloaded, and then
+   * the finalized shuffle merges will be updated.
+   */
+  @VisibleForTesting
+  void reloadAndCleanUpAppShuffleInfo(DB db) throws IOException {
+    logger.info("Reload applications merged shuffle information from DB");
+    reloadActiveAppAttemptsPathInfo(db);
+    reloadFinalizedAppAttemptsShuffleMergeInfo(db);
+  }
+
+  /**
+   * Reload application attempts local paths information.
+   */
+  private void reloadActiveAppAttemptsPathInfo(DB db) throws IOException {
+    if (db != null) {
+      DBIterator itr = db.iterator();
+      itr.seek(APP_ATTEMPT_PATH_KEY_PREFIX.getBytes(StandardCharsets.UTF_8));
+      while (itr.hasNext()) {
+        Map.Entry<byte[], byte[]> e = itr.next();
+        String key = new String(e.getKey(), StandardCharsets.UTF_8);
+        if (!key.startsWith(APP_ATTEMPT_PATH_KEY_PREFIX)) {
+          break;
+        }
+        AppAttemptId appAttemptId = parseDbAppAttemptPathsKey(key);
+        AppPathsInfo appPathsInfo = mapper.readValue(e.getValue(), AppPathsInfo.class);
+        appsShuffleInfo.put(appAttemptId.appId, new AppShuffleInfo(
+            appAttemptId.appId, appAttemptId.attemptId, appPathsInfo));
+      }
+    }
+  }
+
+  /**
+   * Reload the finalized shuffle merges.
+   * Since the deletion of finalized shuffle merges are triggered asynchronously, there can be
+   * cases that deletions miss the execution during restart. Those dangling finalized shuffle merge
+   * keys in the DB will be cleaned up during this reload, if there are no relevant application
+   * attempts local paths information registered in the hashmap.
+   */
+  private void reloadFinalizedAppAttemptsShuffleMergeInfo(DB db) throws IOException {
+    List<byte[]> dbKeysToBeRemoved = new ArrayList<>();
+    if (db != null) {
+      DBIterator itr = db.iterator();
+      itr.seek(APP_ATTEMPT_SHUFFLE_FINALIZE_STATUS_KEY_PREFIX.getBytes(StandardCharsets.UTF_8));
+      while (itr.hasNext()) {
+        Map.Entry<byte[], byte[]> e = itr.next();
+        String key = new String(e.getKey(), StandardCharsets.UTF_8);
+        if (!key.startsWith(APP_ATTEMPT_SHUFFLE_FINALIZE_STATUS_KEY_PREFIX)) {
+          break;
+        }
+        AppAttemptShuffleMergeId partitionId = parseDbAppAttemptShufflePartitionKey(key);
```

Comment:
`partitiionId` could be null

Reply:
I assume this will return null only when the schema gets changed for this key. And in this case, it will throw an exception out and the reload will throw the exception to the caller. The whole initialization will fail if the schema is incompatible. It is not getting handled [here](https://github.com/apache/spark/blob/master/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java#L468) either in the ExternalShuffleBlockResolver.

Reply:
`parseDbAppAttemptShufflePartitionKey` will return null, if there was an exception because one of the other methods it calls was catching the Exception and returns `null`. We need to make that other method not return null and throw exception so the callers can handle it.